### PR TITLE
Updated composite panel

### PR DESCRIPTION
**Pull request by barnettwilliam**
Rebase of PR #134 to incorporate the recent platform changes.

### CODE REVIEW

**Comment by barnettwilliam** [comment]
The core composite panel code has been tested and is functioning in 39375bb59dc10d79a76a9c4bbfddb16319b51130. Using  conversion function `function-flexmi2plantuml` wasn't possible so it was invoked it as an action function via an ActionButtion for testing by modifying the epsilon tool; issue #152 raised to address this. 

ep-composite-panel

File: platform/src/Playground.js [hunk 1]
```diff
@@ -358,7 +375,7 @@ function createPanelForDefinitionId(panel){
             // No activity defined buttons
             newPanel.addButtons( Button.createButtons( panelDefinition.buttons, panel.id));
 
-        } else if (panel.buttons != null && panelDefinition.buttons != null) {
+        } else if (panel.buttons != null && panelDefinition.buttons == null) {
```
Comment (szschaler) [inline platform/src/Playground.js:378]: This change seems unrelated?

File: platform/src/Playground.js [hunk 2]
```diff
@@ -762,7 +774,13 @@ function handleResponseActionFunction(action, requestPromise){
                 switch (action.outputType){
                     case "code":
                         // Text
+                        var editor = outputPanel.getEditor();
```
Comment (szschaler) [inline platform/src/Playground.js:777]: Why do we need to keep this logging code?

**Review by szschaler — CHANGES_REQUESTED** [review]
Quite a large change. Not everything seems related to the core change proposed?

I note further opportunities for refactoring. I am open to deferring some of that refactoring to a later issue, but would like to understand whether it's feasible to do this efficiently now...

File: platform/src/ActivityManager.js [hunk 3]
```diff
@@ -94,6 +94,26 @@ class ActivityManager {
...
+     * 
+     * @param {*} panelRef 
+     * @param {*} panelList 
+     * @returns 
+     */
+    findPanel(panelRef, panelList) {
+        for(let panel of panelList) {
+            if (panel.id == panelRef) return panel;
+            else {
+                if (panel.childPanels) {
```
[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/ActivityManager.js:108]: ActivityManager owns the list of panels. To refactor this code as you mention, we could do as follows:

```
class ActivityManager {
    // ...
    resolvePanelReference(activityId, panelRef){
        if (this.activities[activityId] && this.activities[activityId].panels) {
            const foundPanel = this.activities[activityId].panels.find(pnl => pnl.findPanel(panelRef));
            if (foundPanel !== undefined && typeof foundPanel.id === "string"){
                // error handling logic
            }
        }
    }
    // ...
}

class Panel {
    // ...
    findPanel(panelRef) {
        if (this.id === panelRef) return this;
        else return null;
    }
    // ...
}

class CompositePanel extends Panel {
    // ...
    findPanel(panelRef) {
        return this.childPanels.find(pnl => pnl.findPanel(panelRef));
    }
    // ...
}
```

[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/ActivityManager.js:108]: ActivityManager contains the list of panels. To refactor the code as you mention, Panel could have the findPanel method and CompositePanel would propagate the search to its components:

```
class ActivityManager {
    // ...
    resolvePanelReference(activityId, panelRef){
        if (this.activities[activityId] && this.activities[activityId].panels) {
            const foundPanel = this.activities[activityId].panels.find(pnl => pnl.findPanel(panelRef));
            if (foundPanel !== undefined && typeof foundPanel.id === "string"){
                // Additional logic here
            }
        }
    }
    // ...
}

class Panel {
    // ...
    findPanel(panelRef) {
        if (this.id === panelRef) return this;
        else return null;
    }
    // ...
}

class CompositePanel extends Panel {
    // ...
    findPanel(panelRef) {
        return this.childPanels.find(pnl => pnl.findPanel(panelRef));
    }
    // ...
}
```

File: platform/src/ActivityManager.js [hunk 4]
```diff
@@ -326,6 +346,34 @@ class ActivityManager {
         return null;
     }
 
+
+
+    /* resolve panel refs recursively because of CompositePanels */
+    resolveRef(panelList) {
+        for ( let apanel of panelList ){
```
[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/ActivityManager.js:353]: This could be done following the pattern shown for findPanel.

File: platform/src/Playground.js (hunk 2 above)
Comment (barnettwilliam) [reply platform/src/Playground.js:777]: Agree this should be removed.

File: platform/src/OutputPanel.js [hunk 5]
```diff
@@ -1,9 +1,9 @@
```
[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/OutputPanel.js:1]: ModelPanel and MetamodelPanel can be regarded as specific examples that are generalized by CompositePanel (say, ad-hoc instantiations of the CompositePanel pattern). The rendering of models/metamodels as images has been factored out as actions.

File: platform/src/CompositePanel.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import { Panel } from "./Panel.js";
+import { Layout } from "./Layout.js";
+
+class CompositePanel extends Panel {
```
[the remark this answers is not in the document]
Comment (barnettwilliam) [reply platform/src/CompositePanel.js:4]: Agree this should be added.

File: platform/src/Playground.js [hunk 7]
```diff
@@ -340,6 +344,19 @@ function createPanelForDefinitionId(panel){
 
             break;
 
+            case "CompositePanel":
+
+                newPanel = new CompositePanel(newPanelId);
+                if (panel.childPanels) {
```
[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/Playground.js:350]: Yes, but this also applies to other types of panels in the switch statement.

File: platform/src/Playground.js (hunk 1 above)
Comment (arturboronat) [reply platform/src/Playground.js:378]: A CompositePanel definition may not define buttons

```
 panelDefs:
    - id: composite-panel
      panelclass: CompositePanel
      icon: diagram
```

And these are defined in the activity:

```
 panels:
      - id: panel-yamtl
        name: Transformation(YAMTL)
        ref: yamtl
        file: CD2DB.groovy
      - id: smm-panel-composite
        name: Source Metamodel
        ref: composite-panel
        childrenPanels:
          - id: panel-smm
            name: EMFatic
            ref: emfatic
            file: CD.emf
          - id: panel-smm-diagram
            name: Diagram
            ref: plantuml
        buttons:
          - id: show-editor-button
            icon: editor
            internal: toggle
            targetPanel: panel-smm
            hint: Toggle editor
          - id: show-diagram-button
            icon: diagram
            internal: toggle
            targetPanel: panel-smm-diagram
            hint: Toggle diagram
```

The case where the panelDefinition had no buttons was not considered:

```
        if (panel.buttons == null && panelDefinition.buttons != null){

        } else if (panel.buttons != null && panelDefinition.buttons == null) {
```

Having said that, I have not tested the case `panel.buttons != null && panelDefinition.buttons != null`, where I the panel should override the buttons I guess.

File: platform/src/Playground.js (hunk 2 above)
Comment (arturboronat) [reply platform/src/Playground.js:777]: Not needed. It was used for debugging purposes.

File: platform/src/Playground.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (szschaler) [reply platform/src/Playground.js:350]: Absolutely. I was just pointing out another refactoring opportunity. This one isn't perhaps so specifically driven by the introduction of the new type of Panel, which is why I feel it may not need to be fixed for this PR to be accepted.

**Comment by szschaler** [comment]
Thanks, @arturboronat and @barnettwilliam for discussing these comments so extensively. What's the most efficient way of splitting the work? (I guess, really I'm asking @arturboronat whether you have any capacity of lending a hand with some of these changes?)

File: platform/src/ConsolePanel.js [hunk 8]
```diff
@@ -59,8 +63,8 @@ class ConsolePanel extends Panel {
         root.setAttribute("id", this.id + "Panel");
 
         var editor = document.createElement("div");
-        editor.setAttribute("id", this.id + "Editor");
```
[the remark this answers is not in the document]
Comment (arturboronat) [reply platform/src/ConsolePanel.js:62]: The lines were swapped to keep the symmetry in the code. That is the element properties are now initialized in the same order.

**Comment by arturboronat** [comment]
It's probably easier if I refactor `findPanel` and `resolveRef`. Anything else other than that?

**Comment by szschaler** [comment]
Thanks, @arturboronat, that would be extremely helpful!

**Comment by arturboronat** [comment]
Thank you both for working on this pull request.

I merged the changes to my codebase by there is not much I can do :-)

The proposed refactoring won't work because these functions (findPanel and resolveRef) are used to perform data bindings (converting JSON objects into class instances). These are scripting functions processing JSON/YAML objects and we can't benefit from OOP features like inheritance until we have JS objects.

On the other hand, the activity configuration used in the [YAMTL playground](https://yamtl.github.io/playground/?activities=https://yamtl.github.io/playground-activities/yamtl-demo-activity.yml) fails to pass the validator. See the errors:

Screenshot 2023-12-12 205135

The [OCL activity under platformtools](https://github.com/mdenet/educationplatform-examples/tree/main/ocl) does not have composite panels (it's three months old) and I don't know where [the example above](https://github.com/mdenet/educationplatform/pull/151#issuecomment-1851991983) can be found.

@barnettwilliam Could you test the validator with the YAMTL activity https://yamtl.github.io/playground-activities/yamtl-demo-activity.yml ?

@szschaler Do you see any other way of performing some meaningful refactoring?

**Comment by szschaler** [comment]
@arturboronat ah yes. Looking at this again, I realise my misunderstanding. But that really only means we need to refactor more consistently: it should be the initialisation function of the panel that gets passed in the JSON sub-structure providing the details of this panel and then using that to setup the panel, possibly recursively so. 

Once we have this, all the rest can happen over `Panel` objects and, thus, can be managed through methods within `Panel` that are overwritten in its sub-classes if required.

**Comment by barnettwilliam** [comment]
@arturboronat yes I'll test with your example, it's likely the validation rules need to be updated. I'm surprised the example I tested didn't pick that up. I'll also update the ocl example in mdenet/educationplatform-examples.

**Comment by arturboronat** [comment]
@szschaler The initialize() function is already defined/overriden in each Panel class and the idea of passing a JSON document as parameter would help refactor the switch statement mentioned above but it is not specific to CompositePanel. The search for the relevant JSON document when looking for panel details seems to belong to ActivityManager though.

**Comment by barnettwilliam** [comment]
@arturboronat do you have a version your activity files available for me to test with? they don't appear to be available in the YAMTL repositories. A change I made was to rename childrenPanels to childPanels so the activity needs to be updated to reflect this.

**Comment by arturboronat** [comment]
@barnettwilliam The activity is https://yamtl.github.io/playground-activities/yamtl-demo-activity.yml

In that file the `file:` value for model transformation, metamodels and models can be obtained by prepending the prefix: https://yamtl.github.io/playground-activities/<current value of file>

**Comment by barnettwilliam** [comment]
> @barnettwilliam The activity is https://yamtl.github.io/playground-activities/yamtl-demo-activity.yml
> 
> In that file the `file:` value for model transformation, metamodels and models can be obtained by prepending the prefix: https://yamtl.github.io/playground-activities/. For example, when the value is `CD2DB.groovy` it should be updated to https://yamtl.github.io/playground-activities/CD2DB.groovy

Thanks Artur, when using the activity prefix suggested and  childPanels naming the activity validates functions correctly.

**Comment by arturboronat** [comment]
Fab, I'll try this later today.

**Comment by barnettwilliam** [comment]
> @arturboronat ah yes. Looking at this again, I realise my misunderstanding. But that really only means we need to refactor more consistently: it should be the initialisation function of the panel that gets passed in the JSON sub-structure providing the details of this panel and then using that to setup the panel, possibly recursively so.
> 
> Once we have this, all the rest can happen over `Panel` objects and, thus, can be managed through methods within `Panel` that are overwritten in its sub-classes if required.

@szschaler does this mean that the ActivityManager will then take on the responsibility of creating Panel instances? the approach  I have been taking is to keep ActivityManager and ToolsManager  handling config file objects that are free of any UI code.  Then, have the Playground.js construct the platform level objects like Panel. And if this is the case, do we then favour instantiating objects from configs earlier and mind Panel objects with UI information being included in the ActivityManager etc.?

File: platform/src/CompositePanel.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (barnettwilliam) [reply platform/src/CompositePanel.js:4]: Added implementations of `canSave()` and `save()` in e927029b5dbd71f99e9bbda08ee2c476a3d4be9a, successfully tested multiple saveable panels nested to 3 levels deep.

File: platform/src/Playground.js (hunk 2 above)
Comment (barnettwilliam) [reply platform/src/Playground.js]: Removed debug logs in 92d27f5df4c96eba4476c6af1fb71bf9ad925572

**Comment by barnettwilliam** [comment]
Updated documentation can be found mdenet/educationplatform.wiki@9bb39b88d4c230a6db1653f3e1c1af9837d7ab39  branch `feature/composite-panel`

File: platform/src/CompositePanel.js [hunk 9]
```diff
@@ -15,6 +15,30 @@ class CompositePanel extends Panel {
...
+        this.childPanels.forEach( (cp) => childrenCanSave = childrenCanSave || cp.canSave() );
+
+        return childrenCanSave;
+    }
+
+    /**
+     * @override
+     */
+    save(fileHandler){
+        this.childPanels.forEach( (cp) => {
```
Comment (szschaler) [inline platform/src/CompositePanel.js:35]: This doesn't look right: save needs to return a promise. I was expecting something more along the lines of

```
return Promise.all(this.childPanels.map(p => if (p.canSave()) { return p.save(fileHandler); } else { return null; });
```

**Review by szschaler — CHANGES_REQUESTED** [review]
Happy except for the implementation of `save()`

**Comment by szschaler** [comment]
Thanks for these comments about the proposed refactoring. It looks like this is significantly more complex than I thought and, therefore, really doesn't belong into this PR. I have resolved the corresponding conversations and am happy for this to be merged, subject to a small fix to the implementation of `save{}`.

File: platform/src/CompositePanel.js (hunk 9 above)
Comment (barnettwilliam) [reply platform/src/CompositePanel.js]: Oops yes I didn't notice that it returned a promise -  Fixed as suggested in 844db6729f1a71bab2df9d687c83d266242de68a

**Review by szschaler — APPROVED** [review]
